pandas/core/dtypes/cast.py

```diff
@@ -1225,6 +1225,9 @@ def maybe_cast_to_datetime(
     _ensure_nanosecond_dtype(dtype)
 
     if lib.is_np_dtype(dtype, "m"):
+        if isinstance(value, np.ndarray) and value.ndim == 2 and value.shape[1] == 1:
```

Comment:
Why would e.g. `[None, 1]` be converted to a 2D array?

Reply:
Following the Traceback, the input list is converted to a 2D array in `ndarray_to_mgr`, which returns `_ensure_2d(values)`.
https://github.com/pandas-dev/pandas/blob/9e10119dc8c3ad34cee53e113afedd90cf70a0ec/pandas/core/internals/construction.py#L267
It seems for other types of input, `_ensure_2d` is also called.

Reply:
OK I see. Generally I think there would still be problems if a user passes a nested list (e.g. a "2x2" nested list) or a user passes `dtype="datetime64[unit]"`

I think generally `maybe_cast_to_datetime` should assume the incoming `value` is 1D since the `_from_sequence` calls assume the values are 1D also, so the DataFrame should apply this column column-wise.

Reply:
For nested list, `nested_data_to_arrays` in `DataFrame.__init__` processes the data column-wise, so there is no problem.

There is no error for datetime64, because `DatetimeArray._from_sequence` actually happens to work with 2D array:
```py
from pandas.core.arrays import DatetimeArray, TimedeltaArray
arr = DatetimeArray._from_sequence(np.array([[np.nan], [1]]), dtype="datetime64[ns]")
```

Reply:
I think it is better to move the ravel and reshape in `_try_cast` below, sort of like the Unicode string dtype elif branch, so as to ensure the input of `maybe_cast_to_datetime` is 1D.
https://github.com/pandas-dev/pandas/blob/4651ddb6bca4ee833a22bd280f2ae924813cf50f/pandas/core/construction.py#L796-L810

Reply:
cc @jbrockmendel if you have thoughts on this approach